### PR TITLE
Use custom parsing

### PR DESCRIPTION
**Pull request by pickfire**

### CODE REVIEW

File: src/lib/parse_format.rs [hunk 1]
```diff
@@ -20,45 +10,69 @@ pub enum FormatToken {
...
+        ])
+    );
 
-fn parse_anything(input: &str) -> IResult<&str, FormatToken> {
-    // Parse a single token - either a %-token or a literal.
-    alt((parse_format, parse_literal))(input)
-}
+    assert_eq!(parse_format_string("%g"), Ok(vec![FormatToken::Geometry]));
+
+    assert_eq!(parse_format_string("%-"), Err("Unknown format %-".into()));
```
Comment (tnjd) [inline src/lib/parse_format.rs]: assert!(foo.is_err())

That way we don't test the specific error string.

**Review by Lonami — COMMENTED** [review]
Patch looks fine, although I would've left the changes to `pub` modifiers for another PR.

File: src/lib/parse_format.rs [hunk 2]
```diff
@@ -20,45 +10,69 @@ pub enum FormatToken {
...
+pub(crate) fn parse_format_string(input: &str) -> Result<Format, String> {
+    let mut tokens = Vec::new();
+    let mut input = input.as_bytes();
 
-pub fn parse_format_string(input: &str) -> Result<Format, String> {
-    match parse_all(input) {
-        Ok(("", v)) => Ok(Format(v)),
-        Err(s) => Err(format!("Format string parse error: {:?}", s)),
-        Ok((s, _)) => Err(format!("Format string parse error near \"{}\"", s)),
+    loop {
```
Comment (tnjd) [inline src/lib/parse_format.rs:20]: It would be nice to replace this with a different loop construct, but I'm not sure what.

**Comment by tnjd** [comment]
I like the tests!

File: src/lib/parse_format.rs [hunk 3]
```diff
@@ -20,45 +10,69 @@ pub enum FormatToken {
...
+        let (token, rest) = match input {
+            [b'%', b'i', rest @ ..] => (FormatToken::WindowId, rest),
+            [b'%', b'g', rest @ ..] => (FormatToken::Geometry, rest),
+            [b'%', b'w', rest @ ..] => (FormatToken::Width, rest),
+            [b'%', b'h', rest @ ..] => (FormatToken::Height, rest),
+            [b'%', b'x', rest @ ..] => (FormatToken::X, rest),
+            [b'%', b'y', rest @ ..] => (FormatToken::Y, rest),
+            [b'%', b'%', rest @ ..] => (FormatToken::Literal("%".to_owned()), rest),
+            [b'%', c, ..] => break Err(format!("Unknown format %{}", char::from(*c))),
```
Comment (tnjd) [inline src/lib/parse_format.rs]: Could you put quotes around the "%{}" ?

File: src/lib/parse_format.rs (hunk 1 above)
Comment (pickfire) [reply src/lib/parse_format.rs]: Why not? Do you know when I first got `Unknown format %43` when doing that?

File: src/lib/parse_format.rs (hunk 2 above)
Comment (pickfire) [reply src/lib/parse_format.rs:20]: Yeah, but I had not found a better way.

File: src/lib/parse_format.rs (hunk 3 above)
Comment (pickfire) [reply src/lib/parse_format.rs]: Sure

**Comment by pickfire** [comment]
@Lonami Yes, but these parts I feel lazy to put it into another change. I could send another patch for other `pub`

File: src/lib/parse_format.rs (hunk 1 above)
Comment (tnjd) [reply src/lib/parse_format.rs]: It's not the %- I object to, it's the "Unknown format". Someone rewording that text later should not get a broken test. Basically I see this as testing implementation a bit too much, which makes it fragile. I would be okay with asserting that "%-" appears in the string.

File: src/lib/parse_format.rs [hunk 4]
```diff
@@ -20,45 +10,69 @@ pub enum FormatToken {
...
+    loop {
+        let (token, rest) = match input {
+            [b'%', b'i', rest @ ..] => (FormatToken::WindowId, rest),
+            [b'%', b'g', rest @ ..] => (FormatToken::Geometry, rest),
+            [b'%', b'w', rest @ ..] => (FormatToken::Width, rest),
+            [b'%', b'h', rest @ ..] => (FormatToken::Height, rest),
+            [b'%', b'x', rest @ ..] => (FormatToken::X, rest),
+            [b'%', b'y', rest @ ..] => (FormatToken::Y, rest),
+            [b'%', b'%', rest @ ..] => (FormatToken::Literal("%".to_owned()), rest),
+            [b'%', c, ..] => break Err(format!("Unknown format '%{}'", char::from(*c))),
```
Comment (tnjd) [inline src/lib/parse_format.rs]: This could be "*c as char" instead of "char::from(*c)"

**Comment by tnjd** [comment]
@pickfire I would prefer if you pushed new commits rather than force-pushing

File: src/lib/parse_format.rs (hunk 4 above)
Comment (pickfire) [reply src/lib/parse_format.rs]: Nice, thanks for showing that.

File: src/lib/parse_format.rs (hunk 1 above)
Comment (pickfire) [reply src/lib/parse_format.rs]: Okay, I will change the test case.

**Pull request merged**